### DISCUSSION TITLE
Unble to create a Diart source with a PyAV.AudioFrame as stream.on_next input

### DISCUSSION THREAD

**Discussion by zucher** [juanmc2005/diart#168] [Q&A]
Hi, 
I'm currently creating a service allowing to retrieve the AudioFrame generate by the aiortc webrtc stack, in order to have a live diarization.

I converting the AudioFrame into mono, and  extracting the associated ndarray thanks to:
```python
audio_frame.to_ndarray()
```
the resulting shape is (1,960)

The resulting error is:
**Waveform must have shape (1, samples) but (1, 1, 960) was found**

when I squeeze it:
```python
audio_frame.to_ndarray().squeeze()
```
I have the following error:
**Temporal features must be 2D or 3D**

So I don't know how to pass over this error, someone already try to do some similar stuff ?

Thanks

**Comment by juanmc2005** [comment]
Hi @zucher, could you provide the full stacktrace so I can know where the error is coming from exactly?

**Reply by zucher** [reply]
Hi @juanmc2005 ,

with:
![image](https://github.com/juanmc2005/diart/assets/3034596/f0ef4baf-e3f6-457a-a70b-7cace5be87e7)

for simple to_ndarray():

![image](https://github.com/juanmc2005/diart/assets/3034596/e760b35a-cd7f-4c1c-b9ac-197e6e389ccf)

for squeeze():
![image](https://github.com/juanmc2005/diart/assets/3034596/38c1ba63-b126-439b-bbc4-2752785e4097)

**Reply by zucher** [reply]
@juanmc2005, Do you need additional informations ?

**Reply by juanmc2005** [reply]
@zucher I am unable to reproduce the error. I'd need to know exactly how you're using diart in this context.

Could you provide a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) that I can run to see the problem?

**Reply by zucher** [reply]
Ok I have to do it soon

**Comment by sorgfresser** [comment]
Pretty sure that the `squeeze()` will remove both dimensions with 1.
It might be useful to verify that you do not end up with a shape of `(960)` which is indeed not 2D. Maybe try `squeeze(axis=0)` and see if that fixes it.

**Comment by juanmc2005** [comment]
@zucher I just realized that the dimensions must be inversed in your example.

From `SpeakerSegmentation` docstrings:
```
        Parameters
        ----------
        waveform: TemporalFeatures, shape (samples, channels) or (batch, samples, channels)
```

So you should simply provide an `audio_frame` with shape `(960, 1)` instead of `(1, 960)`.
I'm planning on creating a documentation page soon so this would be easier to debug and work with in the future.

**Comment by zucher** [comment]
Hi @juanmc2005 , thank you for your reply, unfortunatly after swap , I have also an issue:

![image](https://github.com/juanmc2005/diart/assets/3034596/5edde6cb-eb74-40e2-86c6-c3ab8b119ea0)

with
``` python
        audio_frame = av.audio.resampler.AudioResampler(layout="mono").resample(audio_frame)[0]
        self.stream.on_next(audio_frame.to_ndarray().swapaxes(0,1))
```

My code:

Aiortc audio track interception
``` python
import asyncio
from aiortc import MediaStreamTrack
from aiortc.mediastreams import MediaStreamError
from pathlib import Path
import logging.handlers
import logging
import pydub

from diart.blocks.diarization import OnlineSpeakerDiarization
from diart.inference import RealTimeInference
from diart.sinks import RTTMWriter

from .WebRTCAudioSource import WebRTCAudioSource

HERE = Path(__file__).parent

logger = logging.getLogger(__name__)

class AudioDiarization(MediaStreamTrack):
    """
    A audio stream track that only listen.
    """

    kind = "audio"

    def __init__(self, track, channel, transform, event_emitter):
        super().__init__()  # don't forget this!
        self.track = track
        self.transform = transform
        self.ee = event_emitter
        self.channel = channel

        self.diart_init()
        
        self.sound_chunk = pydub.AudioSegment.empty()
        self.silent_count = 0

    @staticmethod
    def create_transformer(track, channel, transform, event_emitter):
        return AudioDiarization(track, channel, transform, event_emitter)
    
    def diart_init( self):
        pipeline = OnlineSpeakerDiarization()
        self.source = WebRTCAudioSource("", 48000)
        inference = RealTimeInference(pipeline, self.source)
        inference.attach_hooks(lambda ann_wav:
                                logger.info(ann_wav[0].to_rttm())
                                )
        prediction = inference()

    async def recv(self):       
        try:
           if self.track.readyState != "live":
                raise MediaStreamError
            
           audio_frame = await self.track.recv()

           self.source.push_audio_frame(audio_frame)

           return audio_frame

        except Exception as e:
            if self.track.readyState == 'ended':
                raise e
            else:
                logger.error(e)
```

WebRTCAudioSource for diart integration
```python
from diart.sources import AudioSource

from typing import Text, Optional, AnyStr, Dict, Any, Union, Tuple
from rx.subject import Subject
from av.frame import Frame
import av
import numpy as np
import pydub

class WebRTCAudioSource(AudioSource):
    def __init__(self, uri: Text, sample_rate: int):
        super().__init__(uri, sample_rate)
        self.stream = Subject()

    @property
    def duration(self) -> Optional[float]:
        """The duration of the stream if known. Defaults to None (unknown duration)."""
        return None

    def read(self):
        """Start reading the source and yielding samples through the stream."""
        pass

    def close(self):
        """Stop reading the source and close all open streams."""
        pass

    def push_audio_frame(self, audio_frame: Frame):

        audio_frame = av.audio.resampler.AudioResampler(layout="mono").resample(audio_frame)[0]
        self.stream.on_next(audio_frame.to_ndarray().swapaxes(0,1))

```
The last line is failing

**Reply by juanmc2005** [reply]
Ok I just verified and audio sources actually push audio with shape `(1, samples)`.
Can you try to push chunks of the same size, and make that size equal to the step of the pipeline?

If that doesn't work, can you try updating to diart v0.8?
